storage: make TestSkipLargeReplicaSnapshot faster by lowering range size

dt opened the pull request:
Fixes #8159

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/8174)

<!-- Reviewable:end -->

dt committed:
storage: make TestSkipLargeReplicaSnapshot faster by lowering range size

Fixes #8159

tamird commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

How long does the test take before and after?

---

Reviewed 2 of 2 files at r1.
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks pending.

---

*[storage/client_bench_test.go, line 43 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8174#-KO5VkKzn3eYdZCnhr7u:-KO5VkKzn3eYdZCnhr7v:-376835498) ([raw file](https://github.com/cockroachdb/cockroach/blob/2d066bb0cf82854889fee0275edb0b39c4b0abe0/storage/client_bench_test.go#L43)):*
> ```Go
> 	const snapSize = 1 << 25 // 32 MiB
> 
> 	fillTestRange(b, rep, snapSize)
> ```

shouldn't this be `rep.GetMaxBytes()` instead of a constant?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8174#-:-KO5WBIxkz74Pcva-Voi:1376094639)*
<!-- Sent from Reviewable.io -->

dt commented:
on my mbp: 17.324s vs 0.602s
allowing for some fixed costs, that looks like about the ~32x reduction I'd expect.

---

Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks pending.

---

*[storage/client_bench_test.go, line 43 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8174#-KO5VkKzn3eYdZCnhr7u:-KO5WS-31CqBn9huyvfs:-1858007695) ([raw file](https://github.com/cockroachdb/cockroach/blob/2d066bb0cf82854889fee0275edb0b39c4b0abe0/storage/client_bench_test.go#L43)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> shouldn't this be `rep.GetMaxBytes()` instead of a constant?
</details>
I don't think this benchmark cares that the range is _full_, just that it has some known amount of data in it when it snapshots.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8174#-:-KO5XAbrMuP5h_QfwS3i:257955646)*
<!-- Sent from Reviewable.io -->

tamird commented:
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks pending.

---

*[storage/client_bench_test.go, line 43 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8174#-KO5VkKzn3eYdZCnhr7u:-KO5YBMaG5dL8GS6SfBA:-1412881336) ([raw file](https://github.com/cockroachdb/cockroach/blob/2d066bb0cf82854889fee0275edb0b39c4b0abe0/storage/client_bench_test.go#L43)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary>
> I don't think this benchmark cares that the range is _full_, just that it has some known amount of data in it when it snapshots.
</details>
I think we'd want this benchmark to indicate a change in the event that our configs change our range sizes.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8174)*
<!-- Sent from Reviewable.io -->

dt commented:
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks pending.

---

*[storage/client_bench_test.go, line 43 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8174#-KO5VkKzn3eYdZCnhr7u:-KO5Yv4NjndkD8248Kmv:534522234) ([raw file](https://github.com/cockroachdb/cockroach/blob/2d066bb0cf82854889fee0275edb0b39c4b0abe0/storage/client_bench_test.go#L43)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> I think we'd want this benchmark to indicate a change in the event that our configs change our range sizes.
</details>
I didn't write it, so I'm not 100% sure of the intent, but since it disables splits and resets the timer after filling the range, I assume it is just benchmarking the `GetSnapshot` call? 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8174)*
<!-- Sent from Reviewable.io -->

tamird commented:
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks pending.

---

*[storage/client_bench_test.go, line 43 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8174#-KO5VkKzn3eYdZCnhr7u:-KO5ZdD9xOKSCx5ujIjs:-1551233263) ([raw file](https://github.com/cockroachdb/cockroach/blob/2d066bb0cf82854889fee0275edb0b39c4b0abe0/storage/client_bench_test.go#L43)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary>
> I didn't write it, so I'm not 100% sure of the intent, but since it disables splits and resets the timer after filling the range, I assume it is just benchmarking the `GetSnapshot` call? 
</details>
I think I wrote it; yes, it was benchmarking GetSnapshot, but I think it should track with range size.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8174)*
<!-- Sent from Reviewable.io -->

dt committed:
storage: always snapshot *full* ranges in BenchmarkReplicaSnapshot

dt commented:
Review status: 1 of 2 files reviewed at latest revision, 1 unresolved discussion, some commit checks pending.

---

*[storage/client_bench_test.go, line 43 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8174#-KO5VkKzn3eYdZCnhr7u:-KO5a7xVlSUe0YLLFo90:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/2d066bb0cf82854889fee0275edb0b39c4b0abe0/storage/client_bench_test.go#L43)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> I think I wrote it; yes, it was benchmarking GetSnapshot, but I think it should track with range size.
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8174)*
<!-- Sent from Reviewable.io -->

tamird commented on storage/client_bench_test.go:
```diff
@@ -38,7 +38,9 @@ func BenchmarkReplicaSnapshot(b *testing.B) {
-	fillTestRange(b, rep, snapSize)
+	snapSize := rep.GetMaxBytes()
+	fillTestRange(b, rep, int(snapSize))
+	b.SetBytes(snapSize)
```
:+1:

dt commented:
bah, I ran stress on this and it is indeed flaky now for some reason :/

tamird commented:
Odd. And it's not flaky on master?

dt commented:
I don't think so (though it is painfully slow to stress on master so maybe I just didn't wait long enough?). It fails in 80-200 runs, with `Snapshot()` succeeding rather than retuning `ErrSnapshotTemporarilyUnavailable`. 

Unfortunately that sentinel error makes it kinda hard to dig into exactly where a particular `ErrSnapshotTemporarilyUnavailable` comes from since we can't wrap them (at least not without changing in upstream etcd/raft too).

dt committed:
storage: test that fillRange actually fills range

dt committed:
storage: actually disable splits in TestSkipLargeReplicaSnapshot

The TestingKnobs are only read during creation of the new store, so they
need to be set before createTestStore.

dt commented:
Flake was because gossip was undoing my MaxSize setting. Limiting via the zone config seems to work. also discovered the disableSplits bool was doing nothing and we were just getting lucky most of the time.

tamird commented:
Awesome, good work. <img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/> 

---

Reviewed 1 of 1 files at r2, 1 of 1 files at r3, 1 of 1 files at r4, 1 of 1 files at r5, 2 of 2 files at r6, 1 of 1 files at r7, 1 of 1 files at r8.
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[storage/replica_raftstorage_test.go, line 37 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8174#-KO6jdy-y06xRpvEWM5R:-KO6jdy-y06xRpvEWM5S:1438620908) ([raw file](https://github.com/cockroachdb/cockroach/blob/d7b5b27fd29225ecedf7afb87cf61be6b187fe35/storage/replica_raftstorage_test.go#L37)):*
> ```Go
> const valSize = 1 << 10 // 1 KiB
> 
> func fillTestRange(t testing.TB, rep *Replica, size int) {
> ```

while you're here, can you make `size` an int64? just noticed due to the type cast in client_bench_test

ah, i see you did that in r6

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8174#-:-KO6kMuhgPLtEWAXl5cw:1386870306)*
<!-- Sent from Reviewable.io -->

dt committed:
config: make the race detector happy even if TestingSetDefaultZoneConfig is called
